core/src/main/java/com/linecorp/armeria/server/logging/structured/StructuredLoggingService.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.logging.structured;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.logging.MessageLogConsumer;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.common.logging.ResponseLog;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerListener;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceConfig;
+import com.linecorp.armeria.server.logging.LogCollectingService;
+
+/**
+ * A decorating service which provides support of structured and optionally externalized request/response
+ * content logging.
+ *
+ * @param <I> the {@link Request} type
+ * @param <O> the {@link Response} type
+ * @param <L> the type of the structured log representation
+ */
+public class StructuredLoggingService<I extends Request, O extends Response, L>
```

Comment:
What do you think about:

- making this class abstract
- moving all methods of `StructuredLogWriter` into this class and removing `StructuredLogWriter`
- replacing `KafkaStructuredLogWriter` with `KafkaStructuredLoggingService`?

This is practically same but it should simplify service construction:

```java
// Before
builder.serviceUnder("/", myService.decorate(StructuredLoggingService.newDecorator(
                     new KafkaStructuredLogWriter<>(...), ...);
// After
builder.serviceUnder("/", myService.decorate(KafkaStructuredLoggingService.newDecorator(..., ...));
```